src/App.js

```diff
@@ -10,24 +10,54 @@ const ALL_PEOPLE = gql`
   }
 `;
 
+const ADD_PERSON = gql`
+  mutation AddPerson($name: String) {
+    addPerson(name: $name) {
+      id
+      name
+    }
+  }
+`;
+
 export default function App() {
+  const [name, setName] = useState('');
   const {
     loading,
-    data
+    data,
+    refetch,
   } = useQuery(ALL_PEOPLE);
 
+  const [addPerson] = useMutation(ADD_PERSON);
+
   return (
     <main>
       <h1>Apollo Client Issue Reproduction</h1>
       <p>
         This application can be used to demonstrate an error in Apollo Client.
       </p>
+      <div className="add-person">
+        <label htmlFor="name">Name</label>
+        <input 
+          type="text" 
+          name="name" 
+          value={name}
+          onChange={evt => setName(evt.target.value)}
+        />
+        <button
+          onClick={() => {
+            addPerson({ variables: { name } }).then(refetch);
```

Comment:
I wonder if instead of using `refetch` here, we should `modify` the cache to inject the new person into the person array. I know it probably doesn't matter which approach we use, since this is a repro template and fetching here isn't technically going over a network since we've mocked the link, but in a real app we probably wouldn't want to trigger `refetch` (since we already have the result of the mutation, and can just make sure it gets added into the person array in the cache without firing another request over the network). We could remove the `refetch` here and update the `useMutation` call to do something like:

```js
const [addPerson] = useMutation(ADD_PERSON, {
  update(cache, { data: { addPerson } }) {
    cache.modify({
      fields: {
        people(existingRefs = []) {
          const newPersonRef = cache.writeFragment({
            data: addPerson,
            fragment: gql`
              fragment NewPerson on Person {
                id
                name
              }
            `
          });
          return [...existingRefs, newPersonRef];
        }
      }
    });
  }
});
```

but this does add a bit more complexity for sure. @benjamn what do you think?

Reply:
I used `refetch` to avoid using `cache.modify`! I know we've been talking about `cache.modify` as "advanced". It didn't seem like the best idea to include it here.

Reply:
Sounds good to me @jcreighton - it definitely increases the complexity of the template. I'm okay with `refetch` - thanks!

Reply:
I think we could use a `merge` function? Or read and write the data for a simpler example? I'm open to not using `refetch`.

Reply:
I vote for avoiding `cache.modify` and using `cache.{read,write}Query` in the mutation `update` function.